### PR TITLE
Switch from Tensorflow to Pytorch

### PR DESCRIPTION
**Pull request by SenWu**
In this PR, we switch the learning part to PyTorch and support Logistic Regression and LSTM.

One thing to be addressed here is the input of training/prediction function needs to be (candidates, features) instead of features.

### CODE REVIEW

File: fonduer/learning/disc_models/rnn/config.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+            break
+
+        new_dir = os.path.split(current_dir)[0]
+        if current_dir == new_dir:
+            logger.debug("Unable to find config file. Using defaults.")
+            break
+        current_dir = new_dir
+        tries += 1
+
+    return config
```
Comment (lukehsiao) [inline fonduer/learning/disc_models/rnn/config.py]: This should just be the same config in utils, right? Shouldn't be duplicated?

Add tests to `tests/utils/test_config.py` to make sure its behaving as expected.

File: fonduer/learning/disc_models/rnn/lstm.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
+import numpy as np
+import torch
+import torch.nn as nn
+from scipy.sparse import issparse
+
+from fonduer.learning.disc_learning import NoiseAwareModel
+from fonduer.learning.disc_models.rnn.config import get_config
```
Comment (lukehsiao) [inline fonduer/learning/disc_models/rnn/lstm.py]: Probably should use utils config. See previous comment.

File: fonduer/learning/disc_models/rnn/utils.py [hunk 3]
```diff
@@ -25,30 +26,51 @@ def len(self):
...
-    c_output = tf.concat(output, 2)
-    batch_size = tf.shape(c_output)[0]
-    max_length = tf.shape(c_output)[1]
-    index = tf.range(0, batch_size) * max_length + (lengths - 1)
-    flat = tf.reshape(c_output, [-1, 2 * dim])
-    return tf.gather(flat, index)
+        return {v: k for k, v in self.d.iteritems()}
+
+
+def mention_to_tokens(mention, token_type="words", lowercase=False):
```
Comment (lukehsiao) [inline fonduer/learning/disc_models/rnn/utils.py]: Docstring?

File: fonduer/learning/disc_models/rnn/config.py (hunk 1 above)
Comment (lukehsiao) [reply fonduer/learning/disc_models/rnn/config.py]: We will also want to update the docs to show these parameters like we do for features.

**Comment by lukehsiao** [comment]
I'm also seeing this error when running `make docs`

```
WARNING: autodoc: failed to import module 'fonduer.learning'; the following exception was raised:
Traceback (most recent call last):
  File "/home/lwhsiao/repos/fonduer/.venv/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 152, in import_module
    __import__(modname)
  File "/home/lwhsiao/repos/fonduer/fonduer/learning/__init__.py", line 1, in <module>
    from fonduer.learning.disc_models.logistic_regression import LogisticRegression
  File "/home/lwhsiao/repos/fonduer/fonduer/learning/disc_models/logistic_regression.py", line 8, in <module>
    from fonduer.learning.disc_learning import NoiseAwareModel
  File "/home/lwhsiao/repos/fonduer/fonduer/learning/disc_learning.py", line 26, in <module>
    class NoiseAwareModel(Classifier, nn.Module):
TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
```

File: .travis.yml [hunk 4]
```diff
@@ -44,20 +50,30 @@ before_install:
 - python --version
 - pip --version
 
+# Install PyTorch for Linux with Python 3.6 and no CUDA
```
[the remark this answers is not in the document]
Comment (SenWu) [reply .travis.yml]: 👍

File: fonduer/learning/disc_learning.py [hunk 5]
```diff
@@ -1,132 +1,84 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
```
[the remark this answers is not in the document]
Comment (SenWu) [reply fonduer/learning/disc_learning.py]: 👍

File: fonduer/learning/disc_learning.py [hunk 6]
```diff
@@ -1,132 +1,84 @@
...
 
 import numpy as np
-import tensorflow as tf
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+
+from .classifier import Classifier
+from .utils import LabelBalancer, reshape_marginals
```
[the remark this answers is not in the document]
Comment (SenWu) [reply fonduer/learning/disc_learning.py]: 👍

File: fonduer/learning/classifier.py [hunk 7]
```diff
@@ -90,6 +92,7 @@ def score(
                 f_beta = 0.0
             return p, r, f_beta
 
+    # TODO: need update, this only works for debugging labeling functions now
     def error_analysis(
```
[the remark this answers is not in the document]
Comment (SenWu) [reply fonduer/learning/classifier.py]: 👍

File: fonduer/learning/disc_learning.py [hunk 8]
```diff
@@ -1,132 +1,84 @@
...
+
 
-from fonduer.learning.classifier import Classifier
-from fonduer.learning.utils import LabelBalancer, reshape_marginals
+def SoftCrossEntropyLoss(input, target):
```
[the remark this answers is not in the document]
Comment (SenWu) [reply fonduer/learning/disc_learning.py]: Done.

File: fonduer/learning/disc_learning.py [hunk 9]
```diff
@@ -213,29 +158,39 @@ def train(
...
+
+        if "host_device" not in self.model_kwargs:
+            self.model_kwargs["host_device"] = "CPU"
+            self.logger.info("Using CPU...")
+        if (
+            self.model_kwargs["host_device"] in self.gpu
+            and not torch.cuda.is_available()
+        ):
+            self.model_kwargs["host_device"] = "CPU"
+            self.logger.info("GPU is not available, switching to CPU...")
```
[the remark this answers is not in the document]
Comment (SenWu) [reply fonduer/learning/disc_learning.py]: Add one for GPU as well.

File: fonduer/learning/disc_models/rnn/config.py (hunk 1 above)
Comment (SenWu) [reply fonduer/learning/disc_models/rnn/config.py]: done!

File: fonduer/learning/disc_models/rnn/lstm.py (hunk 2 above)
Comment (SenWu) [reply fonduer/learning/disc_models/rnn/lstm.py]: done!

File: fonduer/learning/disc_models/rnn/utils.py (hunk 3 above)
Comment (SenWu) [reply fonduer/learning/disc_models/rnn/utils.py]: done!

File: docs/user/learning.rst [hunk 10]
```diff
@@ -21,4 +21,25 @@ insights.
...
 
+Configuration Settings
+----------------------
+
+Visit the `Configuring Fonduer`_ page to see how to provide configuration
+parameters to Fonduer_ via ``.fonduer-config.yaml``.
+
+The different learning parameters are explained in this section.
+
+[TODO] give descriptions for the following::
```
Comment (lukehsiao) [inline docs/user/learning.rst]: Can you help give some descriptions here?

**Pull request merged**